File: WorldFeed/Server/src/Services/Identity/WorldFeed.Identity.API/Client/Clients/AuthClient.cs
namespace WorldFeed.Identity.API.Client.Clients
{
    using System;
    using System.Threading.Tasks;

    using WorldFeed.Common.Public.Parameters.Auth;
    using WorldFeed.Identity.API.Client.Requesters;
    using WorldFeed.Identity.API.Client.Validators;
    using WorldFeed.Identity.API.Infrastructure.Inject;
    using WorldFeed.Identity.API.Models.Authentication;

    public class AuthClient : IAuthClient
    {
        private readonly TwitterClient client;
        private readonly IAuthRequester authRequester;

        public AuthClient(TwitterClient client)
        {
            this.client = client;
            this.authRequester = client.Raw.Auth;
        }

        public IAuthClientParametersValidator ParametersValidator => this.client.ParametersValidator;

        public Task<string> CreateBearerTokenAsync()
        {
            return CreateBearerTokenAsync(new CreateBearerTokenParameters());
        }

        public async Task<string> CreateBearerTokenAsync(ICreateBearerTokenParameters parameters)
        {
            var twitterResult = await this.authRequester.CreateBearerTokenAsync(parameters).ConfigureAwait(false);
            return twitterResult?.Model.AccessToken;
        }

        public async Task InitializeClientBearerTokenAsync()
        {
            var bearerToken = await CreateBearerTokenAsync().ConfigureAwait(false);

            this.client.Credentials = new TwitterCredentials(this.client.Credentials)
            {
                BearerToken = bearerToken
            };
        }

        public Task<IAuthenticationRequest> RequestAuthenticationUrlAsync()
        {
            return RequestAuthenticationUrlAsync(new RequestPinAuthUrlParameters());
        }

        public Task<IAuthenticationRequest> RequestAuthenticationUrlAsync(string callbackUrl)
        {
            return RequestAuthenticationUrlAsync(new RequestUrlAuthUrlParameters(callbackUrl));
        }

        public Task<IAuthenticationRequest> RequestAuthenticationUrlAsync(Uri callbackUri)
        {
            return RequestAuthenticationUrlAsync(new RequestUrlAuthUrlParameters(callbackUri));
        }

        public async Task<IAuthenticationRequest> RequestAuthenticationUrlAsync(IRequestAuthUrlParameters parameters)
        {
            var twitterResult = await this.authRequester.RequestAuthUrlAsync(parameters).ConfigureAwait(false);
            return twitterResult?.Model;
        }

        public async Task<ITwitterCredentials> RequestCredentialsAsync(IRequestCredentialsParameters parameters)
        {
            var twitterResult = await this.authRequester.RequestCredentialsAsync(parameters).ConfigureAwait(false);
            return twitterResult?.Model;
        }

        public Task<ITwitterCredentials> RequestCredentialsFromVerifierCodeAsync(string verifierCode, IAuthenticationRequest authenticationRequest)
        {
            return RequestCredentialsAsync(new RequestCredentialsParameters(verifierCode, authenticationRequest));
        }

        public Task<ITwitterCredentials> RequestCredentialsFromCallbackUrlAsync(string callbackUrl, IAuthenticationRequest authenticationRequest)
        {
            return RequestCredentialsAsync(RequestCredentialsParameters.FromCallbackUrl(callbackUrl, authenticationRequest));
        }

        public Task<ITwitterCredentials> RequestCredentialsFromCallbackUrlAsync(Uri callbackUri, IAuthenticationRequest authenticationRequest)
        {
            return RequestCredentialsAsync(RequestCredentialsParameters.FromCallbackUrl(callbackUri, authenticationRequest));
        }

        public Task<InvalidateTokenResponse> InvalidateBearerTokenAsync()
        {
            return InvalidateBearerTokenAsync(new InvalidateBearerTokenParameters());
        }

        public async Task<InvalidateTokenResponse> InvalidateBearerTokenAsync(IInvalidateBearerTokenParameters parameters)
        {
            var twitterResult = await this.authRequester.InvalidateBearerTokenAsync(parameters).ConfigureAwait(false);
            return twitterResult?.Model;
        }

        public Task<InvalidateTokenResponse> InvalidateAccessTokenAsync()
        {
            return InvalidateAccessTokenAsync(new InvalidateAccessTokenParameters());
        }

        public async Task<InvalidateTokenResponse> InvalidateAccessTokenAsync(IInvalidateAccessTokenParameters parameters)
        {
            var twitterResult = await this.authRequester.InvalidateAccessTokenAsync(parameters).ConfigureAwait(false);
            return twitterResult?.Model;
        }
    }
}

File: Sharebook/Client/src/app/sharebook/Client/Clients/AuthClient.ts
import {IAuthClient, IAuthClientToken} from "../../../core/Public/Client/Clients/IAuthClient";
import {IAuthClientParametersValidator} from "../../../core/Core/Client/Validators/AuthClientParametersValidator";
import {TwitterClient} from "../../TwitterClient";
import {IAuthRequester, IAuthRequesterToken} from "../../../core/Public/Client/Requesters/IAuthRequester";
import {CreateBearerTokenParameters, ICreateBearerTokenParameters} from "../../../core/Public/Parameters/Auth/CreateBearerTokenParameters";
import {ITwitterCredentials, TwitterCredentials} from "../../../core/Public/Models/Authentication/TwitterCredentials";
import {IAuthenticationRequest} from "../../../core/Public/Models/Authentication/IAuthenticationRequest";
import {RequestPinAuthUrlParameters} from "../../../core/Public/Parameters/Auth/RequestPinAuthUrlParameters";
import {IRequestAuthUrlParameters} from "../../../core/Public/Parameters/Auth/IRequestAuthUrlParameters";
import {
  IRequestCredentialsParameters,
  RequestCredentialsParameters
} from "../../../core/Public/Parameters/Auth/RequestCredentialsParameters";
import {InvalidateTokenResponse} from "../../../core/Public/Models/Authentication/InvalidateTokenResponse";
import {
  IInvalidateBearerTokenParameters,
  InvalidateBearerTokenParameters
} from "../../../core/Public/Parameters/Auth/InvalidateBearerTokenParameters";
import {
  IInvalidateAccessTokenParameters,
  InvalidateAccessTokenParameters
} from "../../../core/Public/Parameters/Auth/InvalidateAccessTokenParameters";
import {RequestUrlAuthUrlParameters} from "../../../core/Public/Parameters/Auth/RequestUrlAuthUrlParameters";
import {Inject, Injectable} from "@angular/core";
import {ITwitterClientToken} from "../../../core/Public/ITwitterClient";
import Uri from "typescript-dotnet-commonjs/System/Uri/Uri";

@Injectable({
  providedIn: 'root',
})
export
class AuthClient implements IAuthClient {
  private readonly _client: TwitterClient;
  private readonly _authRequester: IAuthRequester;

  constructor(@Inject(ITwitterClientToken) client: TwitterClient, @Inject(IAuthRequesterToken) auth?: IAuthRequester) {
    this._client = client;
    this._authRequester = auth; // client.raw.auth;
  }

  get parametersValidator(): IAuthClientParametersValidator {
    return this._client.parametersValidator;
  }

  public async createBearerTokenAsync(parameters?: ICreateBearerTokenParameters): Promise<string> {
    let parametersCurrent: ICreateBearerTokenParameters;
    if (!parameters) {
      parametersCurrent = new CreateBearerTokenParameters();
    } else {
      parametersCurrent = parameters;
    }

    let twitterResult = await this._authRequester.createBearerTokenAsync(parametersCurrent); // .ConfigureAwait(false);
    return twitterResult?.model.accessToken;
  }

  public async initializeClientBearerTokenAsync(): Promise<void> {
    let bearerToken = await this.createBearerTokenAsync(); // .ConfigureAwait(false);

    this._client.credentials = new TwitterCredentials(/*this._client.credentials*/);
    this._client.credentials.bearerToken = bearerToken;
  }

  public async requestAuthenticationUrlAsync(callbackUrlOrUriOrParameters?: string | Uri | IRequestAuthUrlParameters): Promise<IAuthenticationRequest> {
    let parameters: IRequestAuthUrlParameters;
    if (callbackUrlOrUriOrParameters) {
      parameters = new RequestPinAuthUrlParameters();
    } else {
      parameters = new RequestUrlAuthUrlParameters(callbackUrlOrUriOrParameters);
    }

    let twitterResult = await this._authRequester.requestAuthUrlAsync(parameters); // .ConfigureAwait(false);
    return twitterResult?.model;
  }


  public async requestCredentialsAsync(parameters: IRequestCredentialsParameters): Promise<ITwitterCredentials> {
    let twitterResult = await this._authRequester.requestCredentialsAsync(parameters); // .ConfigureAwait(false);
    return twitterResult?.model;
  }

  public requestCredentialsFromVerifierCodeAsync(verifierCode: string, authenticationRequest: IAuthenticationRequest): Promise<ITwitterCredentials> {
    return this.requestCredentialsAsync(new RequestCredentialsParameters(verifierCode, authenticationRequest));
  }

  public requestCredentialsFromCallbackUrlAsync(callbackUrlOrUri: string | Uri, authenticationRequest: IAuthenticationRequest): Promise<ITwitterCredentials> {
    return this.requestCredentialsAsync(RequestCredentialsParameters.FromCallbackUrl(callbackUrlOrUri, authenticationRequest));
  }

  public async invalidateBearerTokenAsync(parameters?: IInvalidateBearerTokenParameters): Promise<InvalidateTokenResponse> {
    let parametersCurrent: IInvalidateBearerTokenParameters;
    if (!parameters) {
      parametersCurrent = new InvalidateBearerTokenParameters();
    } else {
      parametersCurrent = parameters;
    }

    let twitterResult = await this._authRequester.invalidateBearerTokenAsync(parametersCurrent); // .ConfigureAwait(false);
    return twitterResult?.model;
  }

  public async invalidateAccessTokenAsync(parameters?: IInvalidateAccessTokenParameters): Promise<InvalidateTokenResponse> {
    let parametersCurrent: IInvalidateAccessTokenParameters;
    if (!parameters) {
      parametersCurrent = new InvalidateAccessTokenParameters();
    } else {
      parametersCurrent = parameters;
    }

    let twitterResult = await this._authRequester.invalidateAccessTokenAsync(parametersCurrent); // .ConfigureAwait(false);
    return twitterResult?.model;
  }
}
